### ISSUE TITLE
FileBodyProducer + LengthEnforcingConsumer cause exceptions in cooperator loop

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7427]
|<img alt="immerrr's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| immerrr reported|
|-|-|
|Trac ID|trac#7427|
|Type|defect|
|Created|2014-05-28 09:08:54Z|

I'm attaching a minimal example that reproduces the described behaviour.  Here's how the stacktrace looks like:

```
Traceback (most recent call last):
Failure: twisted.web._newclient.RequestGenerationFailed: [<twisted.python.failure.Failure <class 'twisted.internet.defer.CancelledError'>>]
Unhandled Error
Traceback (most recent call last):
  File "/home/immerrr/twisted_reproduce.py", line 25, in <module>
    task.react(main)
  File "/home/immerrr/.conda/envs/controlserver/lib/python2.7/site-packages/twisted/internet/task.py", line 858, in react
    _reactor.run()
  File "/home/immerrr/.conda/envs/controlserver/lib/python2.7/site-packages/twisted/internet/base.py", line 1192, in run
    self.mainLoop()
  File "/home/immerrr/.conda/envs/controlserver/lib/python2.7/site-packages/twisted/internet/base.py", line 1201, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/home/immerrr/.conda/envs/controlserver/lib/python2.7/site-packages/twisted/internet/base.py", line 824, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/home/immerrr/.conda/envs/controlserver/lib/python2.7/site-packages/twisted/internet/task.py", line 607, in _tick
    taskObj._oneWorkUnit()
  File "/home/immerrr/.conda/envs/controlserver/lib/python2.7/site-packages/twisted/internet/task.py", line 488, in _oneWorkUnit
    self._completeWith(TaskFailed(), Failure())
  File "/home/immerrr/.conda/envs/controlserver/lib/python2.7/site-packages/twisted/internet/task.py", line 444, in _completeWith
    self._cooperator._removeTask(self)
  File "/home/immerrr/.conda/envs/controlserver/lib/python2.7/site-packages/twisted/internet/task.py", line 580, in _removeTask
    self._tasks.remove(task)
exceptions.ValueError: list.remove(x): x not in list
```

There seem to be two factors causing this:

1. FileBodyProducer doesn't stop producing upon cancellation

2. LengthEnforcingConsumer causes producer to stop twice:

```
LengthEnforcingConsumer.write
|
+-> producer.stopProducing
|   -> task.stop()
|   -> task._completeWith(...)
|
`- raise ExcessWrite
   -> CooperativeTask._oneWorkUnit
   -> except: (catch exception)
   -> task._completeWith(...)

```

Attachments:

* [twisted_reproduce.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c7/9c7f472a7ba7c47e3790268817bb2d45ef428ea7/ca68c3ab7d5b2a11f1229e34e5262346b170f137.py) (661 bytes) - added by immerrr on 2014-05-28 09:09:41Z - minimal example to reproduce the defect

<details><summary>Searchable metadata</summary>

```
trac-id__7427 7427
type__defect defect
reporter__immerrr immerrr
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1401268134000000 1401268134000000
changetime__1446143467497582 1446143467497582
version__None None
owner__ 
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_2"></a>|@wallrj commented|
|-|-|

Hi immerrr

I tried running your script, but didn't see an error with twisted trunk
```
$ python ../../twisted_reproduce.py
```
Please confirm.

**Comment by twisted-trac** [comment]
|<img alt="immerrr's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|immerrr set status to `closed`|
|-|-|

This appears to be fixed somewhere between


```
commit 409d7e66800bb56cb9524e551dc1d8720ccd7185
Author: wsanchez <wsanchez@bbbe8e31-12d6-0310-92fd-ac37d47ddeeb>
Date:   Tue Jun 3 05:20:33 2014 +0000

    Revert r42756: Add feature file and nothing else, whoops.
    
    Committed only the topfile.
    
    Reopens: #6750
    
    
    git-svn-id: svn://svn.twistedmatrix.com/svn/Twisted/trunk@42757 bbbe8e31-12d6-0310-92fd-ac37d47ddeeb

```

and 
```
commit 01b4faf9bf570b8c1904c3fc79e557def6bb7c13
Author: wsanchez <wsanchez@bbbe8e31-12d6-0310-92fd-ac37d47ddeeb>
Date:   Tue Jun 3 05:05:13 2014 +0000

    Merge log-booyah-6750-4: Brief description
    
    Author: wsanchez, glyph
    Reviewer: ralphm
    Fixes: #6750
    
    New logging system.
    
    
    git-svn-id: svn://svn.twistedmatrix.com/svn/Twisted/trunk@42756 bbbe8e31-12d6-0310-92fd-ac37d47ddeeb
```

So, closing this one then.

**Comment by twisted-trac** [comment]
|<img alt="janoonan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|janoonan commented|
|-|-|


Sorry to resurrect an old issue. 

We have an issue in production code with the same stack trace as this ticket. 

For us, it occurs only in load situation, so we are looking through our code to understand if we are doing anything wrong. We do not yet have a reliable way to reproduce the issue. 

However, we have found that the major releases all have the problem described here when the script attached is run. The fix described here seems to have been backed out. 

Could anyone comment if the script shows a real problem or improper usage. (If it's real, I think the ticket should be re-opened).
